components/places/src/bookmark_sync/store.rs

```diff
@@ -223,50 +541,84 @@ impl<'a> BookmarksStore<'a> {
     /// we need to upload, so, if upload is interrupted or fails, we'll stage
     /// the items again on the next sync.
-    fn stage_local_items_to_upload(&self) -> Result<()> {
+    fn stage_items_to_upload(
+        &self,
+        upload_items: &[UploadItem<'_>],
+        upload_tombstones: &[UploadTombstone<'_>],
+    ) -> Result<()> {
+        log::debug!("Cleaning up staged items left from last sync");
+        self.interruptee.err_if_interrupted()?;
+        self.db.execute_batch("DELETE FROM itemsToUpload")?;
+
         // Stage remotely changed items with older local creation dates. These are
         // tracked "weakly": if the upload is interrupted or fails, we won't
         // reupload the record on the next sync.
-        self.db.execute_batch(
-            "INSERT OR IGNORE INTO idsToWeaklyUpload(id)
-             SELECT b.id FROM moz_bookmarks b
-             JOIN mergedTree r ON r.mergedGuid = b.guid
-             JOIN moz_bookmarks_synced v ON v.guid = r.remoteGuid
-             WHERE r.useRemote AND
-                  b.dateAdded < v.dateAdded",
-        )?;
-
-        // Stage remaining locally changed items for upload.
+        log::debug!("Staging items with older local dates added");
+        self.interruptee.err_if_interrupted()?;
         self.db.execute_batch(&format!(
-            "WITH RECURSIVE
-             {local_items_fragment}
-             INSERT INTO itemsToUpload(id, guid, syncChangeCounter, parentGuid,
-                                       parentTitle, dateAdded, title, placeId,
-                                       kind, url, keyword, position)
-             SELECT s.id, s.guid, s.syncChangeCounter, s.parentGuid,
-                    s.parentTitle, s.dateAdded, s.title, s.placeId,
-                    {kind}, h.url, v.keyword, s.position
-             FROM localItems s
-             JOIN mergedTree r ON r.mergedGuid = s.guid
-             LEFT JOIN moz_bookmarks_synced v ON v.guid = r.remoteGuid
-             LEFT JOIN moz_places h ON h.id = s.placeId
-             LEFT JOIN idsToWeaklyUpload w ON w.id = s.id
-             WHERE s.guid <> '{root_guid}' AND
-                   (s.syncChangeCounter > 0 OR w.id NOT NULL)",
-            local_items_fragment = LocalItemsFragment("localItems"),
-            kind = item_kind_fragment("s.type", UrlOrPlaceIdFragment::Url("h.url")),
-            root_guid = BookmarkRootGuid::Root.as_guid().as_str(),
+            "INSERT OR IGNORE INTO itemsToUpload(id, guid, syncChangeCounter,
+                                                 parentGuid, parentTitle, dateAdded,
+                                                 kind, title, placeId, url,
+                                                 keyword, position)
+             {}
+             JOIN itemsToApply n ON n.mergedGuid = b.guid
+             WHERE n.localDateAdded < n.remoteDateAdded",
+            UploadItemsFragment("b", "n.remoteGuid")
```

Comment:
It would be nice to make this fragment more descriptive, something like `UploadItemsFragment("b", RemoteKeywordFragment("n"))`, to make it clearer what `n` is used for.